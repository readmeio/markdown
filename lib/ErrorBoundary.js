import React from 'react';

export default class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return { hasError: true };
  }

  render() {
    if (this.state.hasError) {
      return null;
    }

    // eslint-disable-next-line react/prop-types
    return this.props.children;
  }
}
