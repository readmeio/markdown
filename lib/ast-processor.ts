import type { PluggableList } from 'unified';

import rehypeSlug from 'rehype-slug';
import { remark } from 'remark';
import remarkFrontmatter from 'remark-frontmatter';
import remarkGfm from 'remark-gfm';
import remarkMdx from 'remark-mdx';

import transformers, {
  mermaidTransformer,
  readmeComponentsTransformer,
  variablesTransformer,
  handleMissingComponents,
} from '../processor/transform';

export interface MdastOpts {
  components?: Record<string, string>;
  missingComponents?: 'ignore' | 'throw';
  remarkPlugins?: PluggableList;
}

export const remarkPlugins = [remarkFrontmatter, remarkGfm, ...transformers];
export const rehypePlugins = [rehypeSlug, mermaidTransformer];

const astProcessor = (opts: MdastOpts = {}) => {
  const components = opts.components || {};

  let processor = remark()
    .use(remarkMdx)
    .use(remarkPlugins)
    .use(opts.remarkPlugins)
    .use(variablesTransformer, { asMdx: false })
    .use(readmeComponentsTransformer({ components }));

  if (['ignore', 'throw'].includes(opts.missingComponents)) {
    processor = processor.use(handleMissingComponents, {
      components,
      missingComponents: opts.missingComponents,
    });
  }

  return processor;
};

export default astProcessor;
