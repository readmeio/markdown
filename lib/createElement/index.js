const React = require('react');

const { CodeTabs } = require('../../components/CodeTabs');

const createElement = (type, props, ...children) => {
  const rdmdType = type === 'div' && props?.className === 'code-tabs' ? CodeTabs : type;

  return React.createElement(rdmdType, props, ...children);
};

module.exports = createElement;
