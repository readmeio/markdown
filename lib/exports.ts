import type { MdxjsEsm } from 'mdast-util-mdx';

import { visit } from 'unist-util-visit';

import { isMDXEsm } from '../processor/utils';

import mdast from './mdast';

/* Example mdast structures to find first export name in a mdxjsEsm node:
There are three types of export declarations that we need to consider:
1. VARIABLE DECLARATION
      "type": "mdxjsEsm",
      "value": "export const Foo = () => <div>Hello world</div>\nexport const Bar = () => <div>hello darkness my old friend</div>",
      "data": {
        "estree": {
          "type": "Program",
          "body": [
            {
              "type": "ExportNamedDeclaration",
              "declaration": {
                "type": "VariableDeclaration",
                "declarations": [
                  {
                    "type": "VariableDeclarator",
                    "id": {
                      "type": "Identifier",
                      "name": "Foo" // --------> This is the export name
                    },
                    ...

2/3. FUNCTION DECLARATION & CLASS DECLARATION
      "estree": {
          "type": "Program",
          "body": [
            {
              "type": "ExportNamedDeclaration",
              "declaration": {
                "type": "ClassDeclaration" | "FunctionDeclaration",
                "id": {
                  "type": "Identifier",
                  "name": "Foo" // --------> This is the export name
                },
*/

const exports = (doc: string) => {
  const set = new Set<string>();

  visit(mdast(doc), isMDXEsm, (node: MdxjsEsm) => {
    // Once inside an mdxjsEsm node, we need to check for one or more declared exports within data.estree.body
    // This is because single newlines \n are not considered as a new block, so there may be more than one export statement in a single mdxjsEsm node
    const body = node.data?.estree.body;
    if (!body) return;

    body.forEach(child => {
      if (child.type === 'ExportNamedDeclaration') {
        // There are three types of ExportNamedDeclaration that we need to consider: VariableDeclaration, FunctionDeclaration, ClassDeclaration
        const declaration = child.declaration;

        // FunctionDeclaration and ClassDeclaration have the same structure
        if (declaration.type !== 'VariableDeclaration') {
          // Note: declaration.id.type is always 'Identifier' for FunctionDeclarations and ClassDeclarations
          set.add(declaration.id.name);
        } else {
          declaration.declarations.forEach(({ id }) => {
            if (id.type === 'Identifier') {
              set.add(id.name);
            }
          });
        }
      }
    });
  });

  return Array.from(set);
};

export default exports;
