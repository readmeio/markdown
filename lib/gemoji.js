const { nameToEmoji } = require('gemoji');

const owlmoji = ['owlbert-books', 'owlbert-mask', 'owlbert', 'owlbert-reading', 'owlbert-thinking'];

class Emoji {
  static kind = name => {
    if (name in nameToEmoji) return 'gemoji';
    else if (name.match(/^fa-/)) return 'fontawesome';
    else if (owlmoji.includes(name)) return 'owlmoji';
    return null;
  };

  static nameToEmoji = nameToEmoji;
}

module.exports = Emoji;
