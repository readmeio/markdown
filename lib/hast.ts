import astProcessor, { rehypePlugins, MdastOpts } from './ast-processor';
import remarkRehype from 'remark-rehype';
import { injectComponents } from '../processor/transform';
import { MdastComponents } from '../types';
import mdast from './mdast';

const hast = (text: string, opts: MdastOpts = {}) => {
  const components: MdastComponents = Object.entries(opts.components || {}).reduce((memo, [name, doc]) => {
    memo[name] = mdast(doc);
    return memo;
  }, {});

  const processor = astProcessor(opts).use(injectComponents({ components })).use(remarkRehype).use(rehypePlugins);

  return processor.runSync(processor.parse(text));
};
export default hast;
