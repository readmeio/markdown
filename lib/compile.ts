import { compile as mdxCompile, CompileOptions } from '@mdx-js/mdx';
import remarkFrontmatter from 'remark-frontmatter';
import remarkGfm from 'remark-gfm';

import { defaultTransforms, tailwindRootTransformer } from '../processor/transform';
import { rehypeToc } from '../processor/plugin/toc';
import MdxSyntaxError from '../errors/mdx-syntax-error';
import { rehypePlugins } from './ast-processor';
import { CustomComponents } from '../types';

export type CompileOpts = CompileOptions & {
  lazyImages?: boolean;
  safeMode?: boolean;
  components?: CustomComponents;
  copyButtons?: boolean;
};

const { codeTabsTransformer, ...transforms } = defaultTransforms;

const compile = async (text: string, { components = {}, copyButtons, ...opts }: CompileOpts = {}) => {
  try {
    const vfile = await mdxCompile(text, {
      outputFormat: 'function-body',
      providerImportSource: '#',
      remarkPlugins: [
        remarkFrontmatter,
        remarkGfm,
        ...Object.values(transforms),
        [codeTabsTransformer, { copyButtons }],
        [tailwindRootTransformer, { components }],
      ],
      rehypePlugins: [...rehypePlugins, [rehypeToc, { components }]],
      ...opts,
    });

    return String(vfile).replace('"use strict";', '"use strict";\nconst { React, user } = arguments[0].imports;\n');
  } catch (error) {
    throw error.line ? new MdxSyntaxError(error, text) : error;
  }
};

export default compile;
