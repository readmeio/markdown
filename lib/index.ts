import astProcessor, { MdastOpts, remarkPlugins } from './ast-processor';
import compile from './compile'
import mdx from './mdx';
import run from './run';

export type { MdastOpts };
export { astProcessor, compile, mdx, run, remarkPlugins }
