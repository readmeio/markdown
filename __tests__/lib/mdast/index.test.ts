import { mdast } from '../../../lib';
import { readdir, readFile } from 'node:fs/promises';

describe('mdast transformer', async () => {
  const cases = await Promise.all(
    (await readdir(__dirname))
      .filter(name => !name.match(/.[tj]sx?/))
      .map(async dirname => {
        return {
          name: dirname,
          mdx: await readFile(`${__dirname}/${dirname}/in.mdx`, { encoding: 'utf8' }),
          ast: JSON.parse(await readFile(`${__dirname}/${dirname}/out.json`, { encoding: 'utf8' })),
        };
      }),
  );

  it.each(cases)('parses $name', ({ mdx, ast }) => {
    // @ts-ignore
    expect(mdast(mdx)).toStrictEqualExceptPosition(ast);
  });
});
