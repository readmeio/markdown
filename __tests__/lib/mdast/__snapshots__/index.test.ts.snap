// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mdast transformer > does not remove a component when it is defined in the page and missingComponents === "ignore" 1`] = `
{
  "children": [
    {
      "data": {
        "estree": Node {
          "body": [
            Node {
              "declaration": Node {
                "declarations": [
                  Node {
                    "end": 48,
                    "id": Node {
                      "end": 21,
                      "loc": {
                        "end": {
                          "column": 20,
                          "line": 2,
                          "offset": 21,
                        },
                        "start": {
                          "column": 13,
                          "line": 2,
                          "offset": 14,
                        },
                      },
                      "name": "Inlined",
                      "range": [
                        14,
                        21,
                      ],
                      "start": 14,
                      "type": "Identifier",
                    },
                    "init": Node {
                      "async": false,
                      "body": Node {
                        "children": [
                          Node {
                            "end": 42,
                            "loc": {
                              "end": {
                                "column": 41,
                                "line": 2,
                                "offset": 42,
                              },
                              "start": {
                                "column": 34,
                                "line": 2,
                                "offset": 35,
                              },
                            },
                            "range": [
                              35,
                              42,
                            ],
                            "raw": "Inlined",
                            "start": 35,
                            "type": "JSXText",
                            "value": "Inlined",
                          },
                        ],
                        "closingElement": Node {
                          "end": 48,
                          "loc": {
                            "end": {
                              "column": 47,
                              "line": 2,
                              "offset": 48,
                            },
                            "start": {
                              "column": 41,
                              "line": 2,
                              "offset": 42,
                            },
                          },
                          "name": Node {
                            "end": 47,
                            "loc": {
                              "end": {
                                "column": 46,
                                "line": 2,
                                "offset": 47,
                              },
                              "start": {
                                "column": 43,
                                "line": 2,
                                "offset": 44,
                              },
                            },
                            "name": "div",
                            "range": [
                              44,
                              47,
                            ],
                            "start": 44,
                            "type": "JSXIdentifier",
                          },
                          "range": [
                            42,
                            48,
                          ],
                          "start": 42,
                          "type": "JSXClosingElement",
                        },
                        "end": 48,
                        "loc": {
                          "end": {
                            "column": 47,
                            "line": 2,
                            "offset": 48,
                          },
                          "start": {
                            "column": 29,
                            "line": 2,
                            "offset": 30,
                          },
                        },
                        "openingElement": Node {
                          "attributes": [],
                          "end": 35,
                          "loc": {
                            "end": {
                              "column": 34,
                              "line": 2,
                              "offset": 35,
                            },
                            "start": {
                              "column": 29,
                              "line": 2,
                              "offset": 30,
                            },
                          },
                          "name": Node {
                            "end": 34,
                            "loc": {
                              "end": {
                                "column": 33,
                                "line": 2,
                                "offset": 34,
                              },
                              "start": {
                                "column": 30,
                                "line": 2,
                                "offset": 31,
                              },
                            },
                            "name": "div",
                            "range": [
                              31,
                              34,
                            ],
                            "start": 31,
                            "type": "JSXIdentifier",
                          },
                          "range": [
                            30,
                            35,
                          ],
                          "selfClosing": false,
                          "start": 30,
                          "type": "JSXOpeningElement",
                        },
                        "range": [
                          30,
                          48,
                        ],
                        "start": 30,
                        "type": "JSXElement",
                      },
                      "end": 48,
                      "expression": true,
                      "generator": false,
                      "id": null,
                      "loc": {
                        "end": {
                          "column": 47,
                          "line": 2,
                          "offset": 48,
                        },
                        "start": {
                          "column": 23,
                          "line": 2,
                          "offset": 24,
                        },
                      },
                      "params": [],
                      "range": [
                        24,
                        48,
                      ],
                      "start": 24,
                      "type": "ArrowFunctionExpression",
                    },
                    "loc": {
                      "end": {
                        "column": 47,
                        "line": 2,
                        "offset": 48,
                      },
                      "start": {
                        "column": 13,
                        "line": 2,
                        "offset": 14,
                      },
                    },
                    "range": [
                      14,
                      48,
                    ],
                    "start": 14,
                    "type": "VariableDeclarator",
                  },
                ],
                "end": 49,
                "kind": "const",
                "loc": {
                  "end": {
                    "column": 48,
                    "line": 2,
                    "offset": 49,
                  },
                  "start": {
                    "column": 7,
                    "line": 2,
                    "offset": 8,
                  },
                },
                "range": [
                  8,
                  49,
                ],
                "start": 8,
                "type": "VariableDeclaration",
              },
              "end": 49,
              "loc": {
                "end": {
                  "column": 48,
                  "line": 2,
                  "offset": 49,
                },
                "start": {
                  "column": 0,
                  "line": 2,
                  "offset": 1,
                },
              },
              "range": [
                1,
                49,
              ],
              "source": null,
              "specifiers": [],
              "start": 1,
              "type": "ExportNamedDeclaration",
            },
          ],
          "comments": [],
          "end": 49,
          "loc": {
            "end": {
              "column": 48,
              "line": 2,
              "offset": 49,
            },
            "start": {
              "column": 0,
              "line": 2,
              "offset": 1,
            },
          },
          "range": [
            1,
            49,
          ],
          "sourceType": "module",
          "start": 1,
          "type": "Program",
        },
      },
      "position": {
        "end": {
          "column": 49,
          "line": 2,
          "offset": 49,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "mdxjsEsm",
      "value": "export const Inlined = () => <div>Inlined</div>;",
    },
    {
      "attributes": [],
      "children": [],
      "name": "Inlined",
      "position": {
        "end": {
          "column": 12,
          "line": 4,
          "offset": 62,
        },
        "start": {
          "column": 1,
          "line": 4,
          "offset": 51,
        },
      },
      "type": "mdxJsxFlowElement",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 5,
      "offset": 67,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
